### DISCUSSION TITLE
Why does `AsyncBufReadExt::fill_buf` require `Self: Unpin`?

### DISCUSSION THREAD

**Discussion by Kixunil** [tokio-rs/tokio#5635] [Q&A] [answered]
I'm writing an `async` function that takes a generic `AsyncBufRead`er and repeatedly calls `fill_buf`, processes the bytes and then calls `consume`. I stumbled upon surprising `Unpin` requirement. Why is it needed? It looks to me it should be possible to call `poll_fill_buf` without unpinning. Is this some unfortunate API mistake or is there actual reason I'm missing?

Is it impossible to avoid `Unpin` so I should just require it in my generic function?

**Comment by taiki-e** [comment]
For soundness; see https://github.com/tokio-rs/tokio/pull/2612 for more.

Using `std::pin::pin`, `tokio::pin`, `Box::pin`, etc. in your code may help you avoid adding `Unpin` requirements to your function.

**Comment by Darksonn** [comment, marked as answer]
It's because it takes `&mut self` rather than `Pin<&mut Self>`. You can still use it with `!Unpin` types by using `Self = Pin<&mut T>`.

**Reply by Kixunil** [reply]
> It's because it takes `&mut self` rather than `Pin<&mut Self>`. You can still use it with `!Unpin` types by using `Self = Pin<&mut T>`.

This is what I was missing, thanks!
